torch/_functorch/_activation_checkpointing/rematerialize_nodes_with_ac_annotations_pass.py

```diff
@@ -0,0 +1,130 @@
+"""
+AC rematerialize pass: Duplicates checkpointed nodes for backward, then DCE removes unused forward versions.
+"""
+
+import torch
+import torch.fx as fx
+from torch._functorch import config
+from torch._functorch.compile_utils import raise_getitems
+from torch._functorch.partitioners import (
+    cleanup_recompute_tags,
+    force_save_bw_mutation_src,
+    force_save_collectives,
+    has_recomputable_ops,
+    has_recomputable_rng_ops,
+    is_not_collective,
+    must_recompute,
+)
+
+
+def is_impure_node_for_dce(node):
+    # Check for special collectives that should be treated as pure
+    if not is_not_collective(node):
+        # It's a collective (wait_tensor, all_gather_into_tensor, etc.)
+        # Treat as pure - can be eliminated if unused
+        return False
+
+    # For everything else, fall back to the DEFAULT logic
+    # This is what eliminate_dead_code() calls when is_impure_node=None
+    impure_random = True
+    if torch._guards.TracingContext.try_get():
+        impure_random = torch._inductor.config.fallback_random
+    return node.is_impure(impure_random)
+
+
+def _is_backward_node(node: fx.Node) -> bool:
+    """Check if node is in backward region via annotation"""
+    return node.meta.get("custom", {}).get("remat_pass_tag", None) == "is_backward"
+
+
+def rematerialize_nodes_with_ac_annotations(gm: fx.GraphModule) -> fx.GraphModule:
+    """
+    Duplicate checkpointed nodes for backward use. DCE removes unused forward versions. We assume that
+    you already annotated your backward region with fx.traceback.annotate({"remat_pass_tag": "is_backward"})
+    which helps us identify the backward region.
+    """
+    if not has_recomputable_ops(gm):
+        return gm
+
+    if has_recomputable_rng_ops(gm):
+        raise RuntimeError(
+            "Activation checkpoint rematerializing in `forward-loss-backward` graph does not support RNG ops "
+            "in checkpointed regions. Please move RNG operations outside "
+            "of checkpoint regions, or use joint graph mode (where partitioner handles RNG)."
+        )
+
+    # Use partitioner pass to normalize AC node tags.
+    gm = cleanup_recompute_tags(gm, is_default_partition=False)
```

Comment:
Let's do `is_default_partition=True`

(For context, it mainly is used to control whether the input of AC region is always forced to be saved, which we DO want to do always, but I did not modify the current min-cut behavior for bc reasons.